src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,365 @@
+        self.init_frp();
+        self
+    }
+
+    fn init_frp(&mut self) {
+        let network = &self.network;
+        let frp     = &self.frp;
+        let data    = &self.data;
+
+        // Used to set and detect the end of the tweens. The actual value is irrelevant, only the
+        // duration of the tween matters and that this value is reached after that time.
+        const TWEEN_END_VALUE:f32 = 1.0;
+
+        // Timer used to measure whether the hover has been long enough to show the ports.
+        let delay_show = Tween::new(&network);
+        delay_show.set_duration(SHOW_DELAY_DURATION);
+
+        // Timer used to measure whether the mouse has been gone long enough to hide all ports.
+        let delay_hide = Tween::new(&network);
+        delay_hide.set_duration(HIDE_DELAY_DURATION);
+
+        frp::extend! { network
+
+            // === Size Change Handling == ///
+
+            eval frp.set_size ((size) data.set_size(size.into()));
+
+
+            // === Hover Event Handling == ///
+
+            port_mouse_over      <- source::<PortId>();
+            port_mouse_out       <- source::<PortId>();
+
+            delay_show_finished    <- delay_show.value.map(|t| *t>=TWEEN_END_VALUE );
+            delay_hide_finished    <- delay_hide.value.map(|t| *t>=TWEEN_END_VALUE );
+            on_delay_show_finished <- delay_show_finished.gate(&delay_show_finished).constant(());
+            on_delay_hide_finished <- delay_hide_finished.gate(&delay_hide_finished).constant(());
+
+            visible                <- delay_show_finished.map(|v| *v);
+
+            mouse_over_while_inactive  <- port_mouse_over.gate_not(&visible).constant(());
+            mouse_over_while_active    <- port_mouse_over.gate(&visible).constant(());
+
+            eval mouse_over_while_inactive ([delay_show,delay_hide](_){
+                delay_hide.stop();
+                delay_show.rewind();
+                delay_show.set_end_value(TWEEN_END_VALUE);
+            });
+            eval port_mouse_out ([delay_hide,delay_show](_){
+                delay_show.stop();
+                delay_hide.rewind();
+                delay_hide.set_end_value(TWEEN_END_VALUE);
+            });
+
+            activate_ports <- any(mouse_over_while_active,on_delay_show_finished);
+            eval_ activate_ports (delay_hide.rewind());
+
+            activate_ports_with_selected <- port_mouse_over.sample(&activate_ports);
+
+            hide_all <- on_delay_hide_finished.map(f_!(delay_show.rewind()));
+
+        }
+
+        // Init ports
+        for (index,view) in data.ports.borrow().iter().enumerate() {
+            let shape        = &view.shape;
+            let port_size    = Animation::<f32>::new(&network);
+            let port_opacity = Animation::<f32>::new(&network);
+
+            frp::extend! { network
+
+
+                // === Mouse Event Handling == ///
+
+                eval_ view.events.mouse_over(port_mouse_over.emit(PortId{index}));
```

Comment:
please create smart consturcotr for it. Either in a form of `PortId(index)` or `PortId::new(index)` . you can also use `into`. But using raw constructors everywhere is something we try not to do.